### DISCUSSION TITLE
Instantiating Scala classes from Java

### DISCUSSION THREAD

**Discussion by kiwiidb** [ACINQ/eclair#2629] [Developers]
We succeeded in writing a Java plugin for Eclair that pushes incoming payments to RabbitMQ: https://github.com/getAlby/eclair-plugins . We're also working on a plugin to send boostagrams, but we seem to have some issues when instantiating Scala classes from Java. Both the approaches below result in a NoSuchMethodError
using `new`: 
![Screenshot 2023-04-12 at 08 52 24](https://user-images.githubusercontent.com/33457577/231444997-f14e2a60-0219-403e-bbe6-147029f86cdc.png)
using `apply`: 
![Screenshot 2023-04-12 at 08 52 29](https://user-images.githubusercontent.com/33457577/231445003-92833c47-88f0-4728-b9e9-37dadbb1bae0.png)

Errors:
```
java.lang.NoSuchMethodError: 'void fr.acinq.eclair.payment.send.PaymentInitiator$SendSpontaneousPayment.<init>
(fr.acinq.eclair.MilliSatoshi, fr.acinq.bitcoin.scalacompat.Crypto$PublicKey, fr.acinq.bitcoin.scalacompat.ByteVector32, int, 
scala.Option, fr.acinq.eclair.router.Router$RouteParams, scala.collection.immutable.Seq, boolean)'
```
```
java.lang.NoSuchMethodError: 'fr.acinq.eclair.payment.send.PaymentInitiator$SendSpontaneousPayment 
fr.acinq.eclair.payment.send.PaymentInitiator$SendSpontaneousPayment$.apply(fr.acinq.eclair.MilliSatoshi, 
fr.acinq.bitcoin.scalacompat.Crypto$PublicKey, fr.acinq.bitcoin.scalacompat.ByteVector32, int, scala.Option, 
fr.acinq.eclair.router.Router$RouteParams, scala.collection.immutable.Seq, boolean)'
```
I was told by @t-bast that @sstone might be able to help.

**Comment by sstone** [comment]
Unfortunately Scala is very difficult to use from Java, though I don't see why using `new PaymentInitiator.SendSpontaneousPayment()` wouldn't work.
For example, the following Java code works:

```scala
        Router.SearchBoundaries boundaries = new Router.SearchBoundaries(new MilliSatoshi(1000L), 5.0, 1, new CltvExpiryDelta(10));
        Either<Graph.WeightRatios, Graph.HeuristicsConstants> heuristics = new Left<>(new Graph.WeightRatios(1.0, 0.0, 0.0, 0.0, new Relayer.RelayFees(new MilliSatoshi(100), 100)));
        Router.MultiPartParams mpp = new Router.MultiPartParams(new MilliSatoshi(1000L), 2);
        Router.RouteParams params = new Router.RouteParams(false,
                boundaries,
                heuristics,
                mpp, "", false);
        PaymentInitiator.SendSpontaneousPayment ssp = new PaymentInitiator.SendSpontaneousPayment(
                new MilliSatoshi(1000),
                new Crypto.PublicKey(PublicKey.fromHex("031b84c5567b126440995d3ed5aaba0565d71e1834604819ff9c17f5e9d5dd078f")),
                ByteVector32.One(),
                1,
                Option.apply(""),
                params,
                null,
                false);
```

 It may be a type error (i.e the argument list does not match what is expected), is it possible to share some code or even better add a basic unit test to your project ?

**Reply by kiwiidb** [reply]
Thanks. After looking at it again the problem was fixed when using eclair version 0.8.1-SNAPSHOT as a dependency for the plugin, instead of version 0.8.0.
